### PR TITLE
CV shore point interpolation memory error

### LINKED ISSUE #352: CV Memory error
There's a user that reported a memory error in CV. I was able to reproduce the error below:
https://community.naturalcapitalproject.org/t/coastal-vulnerability-memory-error/1430/9
```python
2020-10-21 13:57:29,379 validation.timeout(638) INFO File checking thread completed.
2020-10-21 13:57:39,231 geos.callback(209) ERROR IllegalArgumentException: Points of LinearRing do not form a closed linestring
2020-10-21 13:57:39,235 coastal_vulnerability._ogr_to_geometry_list(2732) WARNING Could not load feature 95 of C:/Users/ddenu/Workspace/NatCap/User-Data/megc-cv-10-21-2020/InVEST Coastal Vulnerabilty Data-20201021T151429Z-001/InVEST Coastal Vulnerabilty Data/global_polygon.shp
2020-10-21 13:57:40,440 geos.callback(209) ERROR IllegalArgumentException: Points of LinearRing do not form a closed linestring
2020-10-21 13:57:40,441 coastal_vulnerability._ogr_to_geometry_list(2732) WARNING Could not load feature 491 of C:/Users/ddenu/Workspace/NatCap/User-Data/megc-cv-10-21-2020/InVEST Coastal Vulnerabilty Data-20201021T151429Z-001/InVEST Coastal Vulnerabilty Data/global_polygon.shp
2020-10-21 14:05:48,248 geoprocessing.reproject_vector(1640) INFO starting reprojection
2020-10-21 14:05:51,494 geoprocessing.reproject_vector(1683) INFO reprojection 100.0% complete on clipped_projected_landmass.gpkg
2020-10-21 14:05:53,456 coastal_vulnerability.warp_and_mask_bathymetry(1707) INFO Masking positive values from bathymetry
2020-10-21 14:05:56,607 coastal_vulnerability.clip_and_project_raster(2424) WARNING C:/Users/ddenu/Workspace/NatCap/User-Data/megc-cv-10-21-2020/InVEST Coastal Vulnerabilty Data-20201021T151429Z-001/InVEST Coastal Vulnerabilty Data/global_dem.tif is unprojected and will be warped to match the AOI and resampled to a pixel size of 10000 meters
2020-10-21 14:05:58,701 geoprocessing.raster_calculator(375) INFO starting stats_worker
2020-10-21 14:05:58,703 geoprocessing.raster_calculator(381) INFO started stats_worker <Thread(Thread-269, started daemon 22984)>
2020-10-21 14:05:58,922 geoprocessing.raster_calculator(454) INFO 100.0%% complete2020-10-21 15:12:28,466 coastal_vulnerability.prepare_landmass_line_index(643) INFO indexing geometry of landmass
2020-10-21 15:31:41,140 coastal_vulnerability.interpolate_shore_points(674) INFO creating shore points along edge of landmass
2020-10-21 15:39:37,309 Task.add_task(731) ERROR Something went wrong when adding task create shore points (3), terminating taskgraph.
Traceback (most recent call last):
  File "site-packages\taskgraph\Task.py", line 697, in add_task
  File "site-packages\taskgraph\Task.py", line 1174, in _call
  File "site-packages\natcap\invest\coastal_vulnerability.py", line 704, in interpolate_shore_points
  File "site-packages\shapely\ops.py", line 114, in linemerge
  File "site-packages\shapely\geometry\proxy.py", line 30, in _geom
  File "site-packages\shapely\geometry\proxy.py", line 39, in gtag
MemoryError
2020-10-21 15:39:37,311 model._logged_target(1632) ERROR Exception while executing <function execute at 0x18FE5978>
Traceback (most recent call last):
  File "site-packages\natcap\invest\ui\model.py", line 1629, in _logged_target
  File "site-packages\natcap\invest\coastal_vulnerability.py", line 393, in execute
  File "site-packages\taskgraph\Task.py", line 697, in add_task
  File "site-packages\taskgraph\Task.py", line 1174, in _call
  File "site-packages\natcap\invest\coastal_vulnerability.py", line 704, in interpolate_shore_points
  File "site-packages\shapely\ops.py", line 114, in linemerge
  File "site-packages\shapely\geometry\proxy.py", line 30, in _geom
  File "site-packages\shapely\geometry\proxy.py", line 39, in gtag
MemoryError
2020-10-21 15:39:37,313 model._logged_target(1635) INFO Execution finished
2020-10-21 15:39:37,315 utils.prepare_workspace(117) INFO Elapsed time: 1h 42m 8.11999999999989s
2020-10-21 15:39:37,325 execution.run(78) ERROR Target <function InVESTModel.execute_model.<locals>._logged_target at 0x1A54B9C0> failed with exception
Traceback (most recent call last):
  File "site-packages\natcap\invest\ui\execution.py", line 68, in run
  File "site-packages\natcap\invest\ui\model.py", line 1629, in _logged_target
  File "site-packages\natcap\invest\coastal_vulnerability.py", line 393, in execute
  File "site-packages\taskgraph\Task.py", line 697, in add_task
  File "site-packages\taskgraph\Task.py", line 1174, in _call
  File "site-packages\natcap\invest\coastal_vulnerability.py", line 704, in interpolate_shore_points
  File "site-packages\shapely\ops.py", line 114, in linemerge
  File "site-packages\shapely\geometry\proxy.py", line 30, in _geom
  File "site-packages\shapely\geometry\proxy.py", line 39, in gtag
MemoryError
2020-10-21 15:39:37,326 execution.run(83) INFO Execution finished

2020-10-21 14:05:58,923 geoprocessing.raster_calculator(457) INFO signaling stats worker to terminate
2020-10-21 14:05:58,923 geoprocessing.raster_calculator(459) INFO Waiting for raster stats worker result.
2020-10-21 14:05:59,323 coastal_vulnerability.prepare_landmass_line_index(623) INFO preparing landmass geometry
```

### PR DESCRIPTION
**Pull request by dcdenu4**
This PR is a proposed solution for the memory error in `interpolate_shore_points`. The memory error occurs when very large intersecting landmasses are used. The `shapely.ops.linemerge()` call ends up busting memory on these very large landmass line geometries.

The proposed solution uses an arbitrary interval value that is used to "chunk" up how many line geometries make it to the `linemerge` function. In this implementation for each "chunk" shore points are interpolated on that set of lines and written to the vector output. This means that on quite large AOIs and Landmass vectors there will be a slight difference versus interpolating shore points on all of the line geometries. This difference is pretty minimal in reality and I think provides just as good a solution.

The interval value is an added optional parameter to the function.

I did include a test, that uses a very small interval value which gets a different number of shore points interpolated. I'm not sure if this "covers" this new parameter. 

Fixes #352

### CODE REVIEW

**Review by davemfish — CHANGES_REQUESTED** [review]
Hi @dcdenu4 , I have a few minor comments for you to consider. And then I had one larger concern after trying out the test and looking at it's outputs.

I think the test case does cover the new functionality. But there's something strange with point placement. There are two points in the same location, and a larger than expected gap in another location. That will probably be a very minor detail in the context of a "real-world" AOI, but I'd kinda like to understand it and address it. I think it could be a bigger deal if the landmass is very large & complicated (lots of line segments and many times exceeding the linemerge limit) but the model resolution is fairly coarse. Then duplicate points might skew an aggregated result. Or a larger gap could become very noticeable. And that's actually what the data in this test is like: 25m segments on the landmass polygon and a 100m model resolution.

I'd like to give this a little more thought and then maybe chat about it again?

**Comment by dcdenu4** [comment]
Hey @davemfish , sorry about this dumpster fire, good thing we have PRs! I got caught sleeping, programming by coincidence. I'll take a more detailed look over what's happening and your comments. Then we can chat!

**Comment by davemfish** [comment]
> Hey @davemfish , sorry about this dumpster fire, good thing we have PRs! I got caught sleeping, programming by coincidence. I'll take a more detailed look over what's happening and your comments. Then we can chat!

That's not at all the message I was trying to send here! Your PR is exactly what we already discussed. I just hadn't realized until this late stage how I really felt about the solution. And probably couldn't have realized it without seeing it implemented and tested, so thank you for doing that! It's all part of the process. And I'm not even certain if we'll find a nicer solution without expanding the scope of the changes, which I think we're both hesitant to do. But I'll think on that too and then we can talk.

**Comment by dcdenu4** [comment]
@davemfish and I discussed the issues raised above and have decided on a path forward. Keeping this PR alive for the history but going to demote it to Draft.

**Comment by dcdenu4** [comment]
@davemfish okay, tests are passing! 

This PR is updated to combine the interpolation of shore points with the prepping of landmass geometries. The joining of these two processes into one allows for better use of memory management and resources so we don't have to duplicate operations.

**Review by davemfish — CHANGES_REQUESTED** [review]
Hey @dcdenu4 , I think this a huge improvement over the way we used to do this. I think this solution is great, there's just one step in the iterations that I think can be streamlined. See what you think there - about doing the rtree build & the aoi intersection check all in one loop. And some minor comments on the tests. Thanks!

File: src/natcap/invest/coastal_vulnerability.py [hunk 1]
```diff
@@ -690,81 +663,106 @@ def interpolate_shore_points(
...
+            if (line.bounds[0] == line.bounds[2] and
+                    line.bounds[1] == line.bounds[3]):
+                continue
+            polygon_line_rtree.insert(line_id, line.bounds)
+            line_id += 1
+            shapely_line_index.append(line)
+
+        # Already clipped landlines to AOI plus max-fetch extent
+        # still want to clip lines by actual AOI for shore point creation
+        for shapely_landmass_line in shapely_line_index[line_id_previous:]:
```
[the remark this answers is not in the document]
Comment (dcdenu4) [reply src/natcap/invest/coastal_vulnerability.py]: You're totally right about this. Glad I made things more convoluted, would have been to easy to review, haha.

**Comment by dcdenu4** [comment]
@davemfish assuming the new tests pass on GHA I made the changes you suggested here: https://github.com/natcap/invest/pull/370/commits/e0fafc9bb35c393a00e8e91a7978223f77b06487.

**Review by davemfish — APPROVED** [review]
Sweet! All passed. Thanks for merging that test data PR too! @dcdenu4

**Pull request merged**